### ISSUE TITLE
hot() does not forward the ref props to its children

### ISSUE THREAD

**Issue by renchap** [gaearon/react-hot-loader#964]
### Description

When using a `ref` prop in a hot-enabled Root component with `ReactDOM.render`, the parameter passed to the `ref` callback should be the real component, not the proxy component (ref forwarding ?).

Example code:

```es6
var onButtonClick = function() {
  window.helloComponent.alertMessage();
}

class Hello extends React.Component {
  alertMessage() {
    alert(this.props.name);
  }

  render() {
    return (<div> Hello {this.props.name}</div>)
  }
};

ReactDOM.render(
  <Hello name="World" ref={(element) => {window.helloComponent = element}}/>, 
  document.getElementById('container')
);
```

### Expected behavior

When calling `onButtonClick`, `Hello.alertMessage()` should be called

### Actual behavior

What actually happens:

```
Uncaught TypeError: window.helloComponent.alertMessage() is not a function
```

This happens because `window.helloComponent` is a `ProxyComponent` object, not the expected instance of the `Hello` class

### Environment

React Hot Loader version: 4.1.2

**Comment by theKashey** [comment]
Good idea. But only 16.3, you know.

**Comment by renchap** [comment]
React 16.3 adds an easy way to setup ref forwarding, but you can do it manually so it still works with older React versions. The [ref forwarding RFC](https://github.com/reactjs/rfcs/blob/master/text/0030-ref-forwarding.md) gives examples of how it has been implemented before React 16.3.

**Comment by theKashey** [comment]
But it requires a bit different interface.

**Comment by renchap** [comment]
Yes. Maybe implementing it manually using something like `componentRef` would work for now, and switch to a proper forwarded ref once support React < 16.3 is no longer required (probably once React 17 is released)?

**Comment by neoziro** [comment]
I think we could try to add support for "ref" and only for React > 16.3. I don't know if it is possible to do it without breaking React < 16.3 support.

**Comment by theKashey** [comment]
@knoopx - you are right - every component would be wrapped with ProxyComponent, but only `AppContainer` used in `hot` would not forward ref. The rest is more about neutral (except SFC becomes Components)

**Comment by ha404** [comment]
I'm also getting a `ProxyComponent`, this was hard to track down as I went to the documents that don't mention anything about a `ProxyComponent` when dealing with `ref`s. This is breaking my local code :(
